### PR TITLE
New tko io site

### PR DESCRIPTION
**Pull request by brianmhunt**
Consolidate the docs into a new tko.io site

## Summary by CodeRabbit

# Release Notes

* **New Features**
  - Automated documentation deployment to GitHub Pages via GitHub Actions
  - Interactive JSX code examples with live preview in documentation
  - Redesigned documentation site with improved navigation and layout

* **Documentation**
  - Comprehensive TKO API documentation including bindings, observables, and components
  - Added Knockout 3 to 4 migration guide with backwards-compatibility notes

* **Chores**
  - Migrated documentation site to Astro/Starlight framework
  - Updated build tooling and development dependencies

### CODE REVIEW

File: tko.io/src/3to4.md [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+| event `throttle` / `debounce`   | Yes  | Yes  | No  |
+| `ko.Component`   | Yes | Yes  | No  |
+| `applyBindings` returns `Promise`   | Yes  | Yes  | No  |
+| Polyfills included   | None  | Up to ES6  |  `Object.prototype.bind`  |
+| `ko.proxy`   |  Yes  |  No  |  No  |
+| `ko.onError` | Does not automatically throw (you need to add `throw error` to `ko.options.onError`) | Does not rethrow automatically | Always re-throws error |
+| `foreach` binding | FastForEach | Template foreach | Template foreach |
+| `each` binding    | FastForEach | FastForEach | Template foreach
+
+
```
Comment (phillipc) [inline tko.io/src/3to4.md]:
```suggestion
# Security advantages in TKO
TKO is ready for strict Content Security Policies (CSP) and does not use `evil-eval` or `new function` techniques. It also provides HTML sanitization through configurable options. To sanitize HTML templates, override the `options.sanitizeHtmlTemplate` setting using a library such as DOMPurify or Validator.js.
```

File: tko.io/src/content/docs/binding-context/binding-syntax.md [hunk 2]
```diff
@@ -21,7 +26,7 @@ Your value: <input data-bind="value: someValue, valueUpdate: 'afterkeydown'" />
 Cellphone: <input data-bind="value: cellphoneNumber, enable: hasCellphone" />
 ```
 
-The binding *name* should generally match a registered binding handler (either built-in or [custom](custom-bindings.html)) or be a parameter for another binding. If the name matches neither of those, Knockout will ignore it (without any error or warning). So if a binding doesn't appear to work, first check that the name is correct.
+The binding *name* should generally match a registered binding handler (either built-in or [custom](../../binding-context/custom-bindings/)) or be a parameter for another binding. If the name matches neither of those, Knockout will ignore it (without any error or warning). So if a binding doesn't appear to work, first check that the name is correct.
```
[the remark this answers is not in the document]
Comment (phillipc) [reply tko.io/src/content/docs/binding-context/binding-syntax.md:28]: I think the path is correct

File: tko.io/public/js/examples.js [hunk 3]
```diff
@@ -0,0 +1,146 @@
+import * as esbuild from 'https://cdn.jsdelivr.net/npm/esbuild-wasm@0.24.0/esm/browser.min.js';
```
Comment (phillipc) [inline tko.io/public/js/examples.js:1]: Should we port this also to TypeScript?

File: tko.io/.eleventy.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
+import markdownIt from "markdown-it";
```
Comment (phillipc) [inline tko.io/.eleventy.js]: Port to TypeScript at some point?

**Comment by phillipc** [comment]
The AI ​​comments are good. I only left a few small comments.

File: tko.io/src/css/tko.css [hunk 5]
```diff
@@ -0,0 +1,659 @@
...
+  -webkit-font-smoothing: antialiased;
+  -moz-osx-font-smoothing: grayscale;
+}
+
+html,
+body {
+  margin: 0;
+  padding: 0;
+  overflow-x: hidden;
+  font-family: 'Inter', -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen', 'Ubuntu', 'Cantarell', sans-serif;
```
Comment (phillipc) [inline tko.io/src/css/tko.css]: Or Simply, CDN free and the same font everywhere

```suggestion
  font-family: sans-serif;
```

File: tko.io/src/css/tko.css [hunk 6]
```diff
@@ -0,0 +1,659 @@
+@import url('https://fonts.googleapis.com/css2?family=Inter:wght@400;500;600;700&family=JetBrains+Mono:wght@400;500&display=swap');
```
Comment (phillipc) [inline tko.io/src/css/tko.css]:
```suggestion

```
See below

File: tko.io/.eleventy.js (hunk 4 above)
Comment (brianmhunt) [reply tko.io/.eleventy.js]: ideally yes, good point, but I'm not sure how / whether eleventy can import it.

File: tko.io/public/js/examples.js (hunk 3 above)
Comment (brianmhunt) [reply tko.io/public/js/examples.js:1]: at some point, definitely

File: tko.io/src/content/docs/3to4.md [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+| Browser compatibility |  Modern browsers (i.e. [`Proxy`](https://caniuse.com/#feat=proxy) support) | Internet Explorer 9+ |  Internet Explorer 6+  |
+| `ko.BindingHandler`    | Yes  | Yes  | No  |
+| `observableArray.length`   | Yes  | Browsers with [Array subclass](https://kangax.github.io/compat-table/es6/#test-Array_is_subclassable_length_property_(accessing)) support [More info](http://perfectionkills.com/how-ecmascript-5-still-does-not-allow-to-subclass-an-array/#wrappers_prototype_chain_injection)  | No  |
+| iterable `observableArray`   | Yes  | Browsers with [Symbol support](https://kangax.github.io/compat-table/es6/#test-Symbol)  | No  |
+| event `throttle` / `debounce`   | Yes  | Yes  | No  |
+| `ko.Component`   | Yes | Yes  | No  |
+| `applyBindings` returns `Promise`   | Yes  | Yes  | No  |
+| Polyfills included   | None  | Up to ES6  |  `Object.prototype.bind`  |
+| `ko.proxy`   |  Yes  |  No  |  No  |
+| `ko.onError` | Does not automatically throw (you need to add `throw error` to `ko.options.onError`) | Does not rethrow automatically | Always re-throws error |
```
Comment (phillipc) [inline tko.io/src/content/docs/3to4.md:27]: OnError is discussable. In some places, it must throw an exception; otherwise, null pointers or other errors will occur. I have add an Boolean in onError, to identify such cases. @brianmhunt What is the original indentation at this point?

File: tko.io/.eleventy.js (hunk 4 above)
Comment (phillipc) [reply tko.io/.eleventy.js]: It's not necessary: https://www.11ty.dev/docs/languages/typescript/

File: tko.io/src/content/docs/3to4.md [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+
+Knockout 3 used Jasmine 1 to perform tests.  Some packages in TKO continue to use Jasmine 1, while others now use Mocha.
+
+## Backwards compatibility and 💥 breaking changes
+
+We've strived to have backwards compatibility, and the numerous test cases we've run it through have been successful.  A key value of Knockout has been and remains the comprehensive test cases that enforce performance to a predictable standard, and TKO meets all the tests that were enforced on Knockout 3.x, notwithstanding a few issues:
+
+- The `function` no longer works in `data-bind`, having been replaced by lambdas, so `data-bind='click: function () { x++ }'` will have to be replaced by e.g. `data-bind='click: => x++'`
+- The `data-bind` is now interpreted by a LL compiler (based on [knockout-secure-binding](https://github.com/brianmhunt/knockout-secure-binding)), so some edge case parsing issues may crop up
+
```
Comment (phillipc) [inline tko.io/src/content/docs/3to4.md:51]:
```suggestion
- The `descendantsComplete` Binding doesn't support the `completeOn: 'render'`-Option in with-bindings. Instead you can use `childrenComplete` or the promise return from `applyBindings`.

```
@brianmhunt Was that done by design? For me okay

Reference:
https://github.com/knockout/knockout/blob/45e86bf0347cf742ca359bab051ea82fdfae7a5d/src/binding/defaultBindings/ifIfnotWith.js#L15

https://knockoutjs.com/documentation/binding-lifecycle-events.html

File: tko.io/src/content/docs/3to4.md [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+# Security advantages in TKO
+TKO is ready for strict Content Security Policies (CSP) and does not use `evil-eval` or `new function` techniques. It also provides HTML sanitization through configurable options. To sanitize HTML templates, override the `options.sanitizeHtmlTemplate` setting using a library such as DOMPurify or Validator.js.
+# Build System
+
+TKO has a complete build system rewrite and a conversion of the code from a concatenation of source files to discrete ES6 modules with `import` and `export` statements.  This process has been complex and taken years, with the result being a Knockout with a strong foundation of discrete reusable components.
+
+Knockout 4 and TKO will be maintained at the new monorepo [knockout/tko](https://github.com/knockout/tko) on GitHub.  Knockout 3.x will continue to be maintained at [knockout/knockout](https://github.com/knockout/knockout).
+
+TKO now uses standard tools like `lerna` and `yarn`, and employs [StandardJS](https://standardjs.com/) for code style — though all the old code is not yet updated to StandardJS yet.
```
Comment (phillipc) [inline tko.io/src/content/docs/3to4.md:39]:
```suggestion
TKO now uses standard tools like `lerna`, `npm` and `esbuild`, and employs [StandardJS](https://standardjs.com/) for code style — though the complete code base via `prettier`.
```

File: tko.io/src/content/docs/3to4.md [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+
+
+# Version Comparison
+
+| Version | tko.js  | Knockout 4 | Knockout 3.x |
+| --- | --- | --- | --- |
+| Browser compatibility |  Modern browsers (i.e. [`Proxy`](https://caniuse.com/#feat=proxy) support) | Internet Explorer 9+ |  Internet Explorer 6+  |
+| `ko.BindingHandler`    | Yes  | Yes  | No  |
+| `observableArray.length`   | Yes  | Browsers with [Array subclass](https://kangax.github.io/compat-table/es6/#test-Array_is_subclassable_length_property_(accessing)) support [More info](http://perfectionkills.com/how-ecmascript-5-still-does-not-allow-to-subclass-an-array/#wrappers_prototype_chain_injection)  | No  |
+| iterable `observableArray`   | Yes  | Browsers with [Symbol support](https://kangax.github.io/compat-table/es6/#test-Symbol)  | No  |
```
Comment (phillipc) [inline tko.io/src/content/docs/3to4.md:21]: Dead link

File: tko.io/src/content/docs/3to4.md [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+## Backwards compatibility and 💥 breaking changes
+
+We've strived to have backwards compatibility, and the numerous test cases we've run it through have been successful.  A key value of Knockout has been and remains the comprehensive test cases that enforce performance to a predictable standard, and TKO meets all the tests that were enforced on Knockout 3.x, notwithstanding a few issues:
+
+- The `function` no longer works in `data-bind`, having been replaced by lambdas, so `data-bind='click: function () { x++ }'` will have to be replaced by e.g. `data-bind='click: => x++'`
+- The `data-bind` is now interpreted by a LL compiler (based on [knockout-secure-binding](https://github.com/brianmhunt/knockout-secure-binding)), so some edge case parsing issues may crop up
+
+## Browser support
+
+Knockout 4 and TKO and the underlying packages will start support from IE9 forward.  We aim to test Knockout across all browsers from this point forward.
```
Comment (phillipc) [inline tko.io/src/content/docs/3to4.md:54]:
```suggestion
Knockout 4 and TKO and the underlying packages will start support from IE11 forward.  We aim to test Knockout across all browsers from this point forward.
```

File: tko.io/src/content/docs/3to4.md [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+
+If you need IE6, 7, and 8 support you may prefer to keep using the Knockout 3.x line, or be prepared for workarounds or missing functionality.
```
Comment (phillipc) [inline tko.io/src/content/docs/3to4.md:56]:
```suggestion
If you need IE6 to IE10 support you may prefer to keep using the Knockout 3.x line.
```
